core/lib/web3_decl/src/client/mock.rs

```diff
@@ -1,73 +1,202 @@
 //! Mock L2 client implementation.
 
-use std::{fmt, future::Future, pin::Pin, sync::Arc};
+use std::{any, collections::HashMap, fmt, future::Future, marker::PhantomData, sync::Arc};
 
 use async_trait::async_trait;
 use futures::future;
-use jsonrpsee::core::{
-    client::{BatchResponse, ClientT, Error, Subscription, SubscriptionClientT},
-    params::BatchRequestBuilder,
-    traits::ToRpcParams,
+use jsonrpsee::{
+    core::{
+        client::{BatchResponse, ClientT, Error},
+        params::BatchRequestBuilder,
+        traits::ToRpcParams,
+    },
+    types::{error::ErrorCode, ErrorObject},
 };
-use serde::de::DeserializeOwned;
+use serde::{de::DeserializeOwned, Serialize};
 
-use super::TaggedClient;
+use super::{boxed::RawParams, ForNetwork, Network, TaggedClient};
 
-type MockHandleResult<'a> =
-    Pin<Box<dyn Future<Output = Result<serde_json::Value, Error>> + Send + 'a>>;
-type RequestHandler = dyn Fn(&str, serde_json::Value) -> MockHandleResult<'_> + Send + Sync;
+/// Object-safe counterpart to [`Handler`]. We need it because async closures aren't available on stable Rust.
+#[async_trait]
+trait HandleGenericRequest: Send + Sync + fmt::Debug {
+    async fn handle_generic_request(
+        &self,
+        req: serde_json::Value,
+    ) -> Result<serde_json::Value, Error>;
+}
+
+/// The only implementation of [`HandleGenericRequest`].
+struct RequestHandler<Req, H, const ASYNC: bool> {
+    inner: H,
+    _fn: PhantomData<fn(Req)>,
+}
+
+impl<Req, H, const ASYNC: bool> fmt::Debug for RequestHandler<Req, H, ASYNC>
+where
+    Req: 'static,
+    H: Handler<Req, ASYNC>,
+    H::Resp: 'static,
+{
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        formatter
+            .debug_struct("RequestHandler")
+            .field("async", &ASYNC)
+            .field("req", &any::type_name::<Req>())
+            .field("resp", &any::type_name::<H::Resp>())
+            .finish()
+    }
+}
+
+#[async_trait]
+impl<Req, H, const ASYNC: bool> HandleGenericRequest for RequestHandler<Req, H, ASYNC>
+where
+    Req: DeserializeOwned + Send + 'static,
+    H: Handler<Req, ASYNC>,
+    H::Resp: Serialize + 'static,
+{
+    async fn handle_generic_request(
+        &self,
+        req: serde_json::Value,
+    ) -> Result<serde_json::Value, Error> {
+        let req: Req = serde_json::from_value(req)?;
+        let resp = self.inner.call(req).await?;
+        Ok(serde_json::to_value(resp)?)
+    }
+}
+
+/// Request handler for [`MockClient`]. Implemented automatically for sync and async closures
+/// taking zero to 3 de-serializable args (you might want to specify their types explicitly)
+/// and returning `Result<Resp, Error>`, where `Resp` is serializable.
+pub trait Handler<Req, const ASYNC: bool>: Send + Sync + 'static {
```

Comment:
Maybe a bit of an overkill, but it allows to define mocks really easily and fluently. (I may introduce a similar approach for mocking contracts.) Can extract in a separate PR if it makes sense. Inspired [by `axum`](https://docs.rs/axum/latest/axum/handler/trait.Handler.html) (I think other HTTP frameworks use similar handler extraction as well).

Reply:
I am (as always) a bit grumpy when it comes to increasing cognitive complexity, but this change does seem to be useful. Given that it's implemented already, let's keep it here.